### PR TITLE
[ADD] cmis_stock_production_lot

### PR DESCRIPTION
**Pull request by mgosai**
# CMIS Stock Production Lot

This module allows you to upload and display documents to a CMIS folder on a Lot/Serial Number.

# Installation

You need to install `libcmis`.

# Configuration

Before configuring this module, you need to configure Alfresco. Follow the instructions detailed here: http://alfodoo.org/guides/install.html

Then
 * Go to Users > Users > Access Rights > Document Manager (check it True)

# Usage

To use this module:

* Activate the traceability with lot/SN
* Go to Inventory > Master Data > Lot/Serial Numbers
* Create or Select a Lot
* Edit the Lot and go to the tab "Documents" to upload a file

### CODE REVIEW

File: cmis_mrp_lot/models/mrp_lot.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            ret['cmis:secondaryObjectTypeIds'] = \
+                ['P:ursa:documentClassifiable']
+            # Link Partner Document Type field value to Alfresco object
+            ret['ursa:documentType'] = MAP_DOCUMENT_TYPES['document']
+            # Link Partner Created By user field value to Alfresco object
+            ret['ursa:createdBy'] = record.create_uid.partner_id.name or ''
+            # Link Partner Last Modified By user field value to Alfresco object
+            ret['ursa:lastModifiedBy'] = record.write_uid.partner_id.name or ''
+            return_values[record.id] = ret
+        return return_values
```
Comment (max3903) [inline cmis_mrp_lot/models/mrp_lot.py]: @mgosai No need for properties

File: cmis/__manifest__.py [hunk 2]
```diff
@@ -15,8 +15,13 @@
     'external_dependencies': {
         'python': ['cmislib'],
     },
+    'depends': [
+        'base',
+    ],
```
Comment (max3903) [inline cmis/__manifest__.py]: @mgosai No need for this dependency.

**Review by max3903 — CHANGES_REQUESTED** [review]

**Comment by mgosai** [comment]
@max3903 
Updated changes. Please review
Thanks

**Comment by max3903** [comment]
@mgosai Also we'll need to rename the module to `cmis_stock_production_lot`

**Comment by mgosai** [comment]
@max3903 
Renamed Module!
Thanks

**Review by max3903 — APPROVED** [review]

**Comment by max3903** [comment]
@mgosai For some reason, Odoo is trying to create the folder for each lot/SN outside of the specific folder configured on the CMIS backend. This generates a permission denied error:

```
Traceback (most recent call last):
  File "/opt/vsp2/env/lib/python2.7/site-packages/odoo-10.0-py2.7.egg/odoo/http.py", line 641, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/opt/vsp2/env/lib/python2.7/site-packages/odoo-10.0-py2.7.egg/odoo/http.py", line 683, in dispatch
    result = self._call_function(**self.params)
  File "/opt/vsp2/env/lib/python2.7/site-packages/odoo-10.0-py2.7.egg/odoo/http.py", line 333, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/opt/vsp2/env/lib/python2.7/site-packages/odoo-10.0-py2.7.egg/odoo/service/model.py", line 101, in wrapper
    return f(dbname, *args, **kwargs)
  File "/opt/vsp2/env/lib/python2.7/site-packages/odoo-10.0-py2.7.egg/odoo/http.py", line 326, in checked_call
    result = self.endpoint(*a, **kw)
  File "/opt/vsp2/env/lib/python2.7/site-packages/odoo-10.0-py2.7.egg/odoo/http.py", line 941, in __call__
    return self.method(*args, **kw)
  File "/opt/vsp2/env/lib/python2.7/site-packages/odoo-10.0-py2.7.egg/odoo/http.py", line 506, in response_wrap
    response = f(*args, **kw)
  File "/opt/vsp2/alfodoo/cmis_field/controllers/main.py", line 13, in create_field_value
    model_inst._fields[field_name].create_value(model_inst)
  File "/opt/vsp2/alfodoo/cmis_field/fields/cmis_folder.py", line 111, in create_value
    self._create_in_cmis(records, backend)
  File "/opt/vsp2/alfodoo/cmis_field/fields/cmis_folder.py", line 128, in _create_in_cmis
    parent, name, props)
  File "/opt/vsp2/env/local/lib/python2.7/site-packages/cmislib/browser/binding.py", line 1425, in createFolder
    return parentFolder.createFolder(name, properties, **kwargs)
  File "/opt/vsp2/env/local/lib/python2.7/site-packages/cmislib/browser/binding.py", line 2102, in createFolder
    **kwargs)
  File "/opt/vsp2/env/local/lib/python2.7/site-packages/cmislib/browser/binding.py", line 107, in post
    self._processCommonErrors(resp, url)
  File "/opt/vsp2/env/local/lib/python2.7/site-packages/cmislib/cmis_services.py", line 59, in _processCommonErrors
    raise PermissionDeniedException(error['status'], url)
PermissionDeniedException: Error 403 at http://*********************/alfresco/api/-default-/public/cmis/versions/1.1/browser/root
```


screenshot from 2018-08-01 23-29-23


I also get this error when I try to modify the location on a CMIS backend:
```
Traceback (most recent call last):
  File "/opt/vsp2/env/lib/python2.7/site-packages/odoo-10.0-py2.7.egg/odoo/http.py", line 641, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/opt/vsp2/env/lib/python2.7/site-packages/odoo-10.0-py2.7.egg/odoo/http.py", line 683, in dispatch
    result = self._call_function(**self.params)
  File "/opt/vsp2/env/lib/python2.7/site-packages/odoo-10.0-py2.7.egg/odoo/http.py", line 333, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/opt/vsp2/env/lib/python2.7/site-packages/odoo-10.0-py2.7.egg/odoo/service/model.py", line 101, in wrapper
    return f(dbname, *args, **kwargs)
  File "/opt/vsp2/env/lib/python2.7/site-packages/odoo-10.0-py2.7.egg/odoo/http.py", line 326, in checked_call
    result = self.endpoint(*a, **kw)
  File "/opt/vsp2/env/lib/python2.7/site-packages/odoo-10.0-py2.7.egg/odoo/http.py", line 941, in __call__
    return self.method(*args, **kw)
  File "/opt/vsp2/env/lib/python2.7/site-packages/odoo-10.0-py2.7.egg/odoo/http.py", line 506, in response_wrap
    response = f(*args, **kw)
  File "/opt/vsp2/odoo/addons/web/controllers/main.py", line 890, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/opt/vsp2/odoo/addons/web/controllers/main.py", line 882, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/opt/vsp2/env/lib/python2.7/site-packages/odoo-10.0-py2.7.egg/odoo/api.py", line 689, in call_kw
    return call_kw_multi(method, model, args, kwargs)
  File "/opt/vsp2/env/lib/python2.7/site-packages/odoo-10.0-py2.7.egg/odoo/api.py", line 680, in call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/opt/vsp2/alfodoo/cmis_web_proxy/models/cmis_backend.py", line 76, in write
    self.get_proxy_info_by_id.clear_cache(self)
AttributeError: 'function' object has no attribute 'clear_cache'
```

File: cmis_stock_production_lot/models/stock_production_lot.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    @api.multi
+    def _cmis_create_parent_get(self, field, backend):
+        ret = {}
+        # Split path based on subfolder
+        if backend.initial_directory_write.startswith('/'):
+            path_parts = backend.initial_directory_write.split("/")[1:]
+        else:
+            path_parts = backend.initial_directory_write.split("/")
+        # Append model name following path
+        path_parts.append('stock.production.lot')
```
Comment (lmignon) [inline cmis_stock_production_lot/models/stock_production_lot.py]: s/'stock.production.lot'/self._name

File: cmis_stock_production_lot/README.rst [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+=========================
+CMIS Stock Production Lot
+=========================
+
+This module allows you to upload and display documents to a CMIS folder on a Lot/Serial Number.
+
+Installation
+============
+
+You need to install `libcmis`.
```
Comment (lmignon) [inline cmis_stock_production_lot/README.rst]: s/libcmis/cmislib

File: cmis_stock_production_lot/README.rst [hunk 5]
```diff
@@ -0,0 +1,79 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
```
Comment (lmignon) [inline cmis_stock_production_lot/README.rst]: you should use the new way to write the README files https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

**Review by lmignon — COMMENTED** [review]
@mgosai @max3903 Thank you for this module. It's a practical sample on how to integrate a cmis_folder into an existing model.

File: cmis_stock_production_lot/models/stock_production_lot.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    @api.multi
+    def _cmis_create_name_get(self, field, backend):
+        ret = {}
+        for record in self:
+            ret[record.id] = record.name or ''
+        return ret
+
+    @api.multi
+    def _cmis_create_parent_get(self, field, backend):
+        ret = {}
```
Comment (lmignon) [inline cmis_stock_production_lot/models/stock_production_lot.py]: This method implements the default behaviour provided by the cmis_folder field when you don't specify the 'create_parent_get' parameter https://github.com/acsone/alfodoo/blob/10.0/cmis_field/fields/cmis_folder.py#L33.  
You should remove it.

File: cmis_stock_production_lot/models/stock_production_lot.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+    cmis_folder = fields.CmisFolder(
+        copy=False,
+        create_name_get='_cmis_create_name_get',
+        create_parent_get='_cmis_create_parent_get',
+        oldname='cmis_objectid')
+
+    @api.multi
+    def _cmis_create_name_get(self, field, backend):
+        ret = {}
```
Comment (lmignon) [inline cmis_stock_production_lot/models/stock_production_lot.py]: IMO, This method will return the same result as the default one provided by the cmis_folder field when you don't specify the 'create_name_get' parameter https://github.com/acsone/alfodoo/blob/10.0/cmis_field/fields/cmis_folder.py#L44. You should remove it
You can remove it.

**Comment by max3903** [comment]
@lmignon Can you update your review please?

**Review by smangukiya — APPROVED** [review]
Changes seems good to me.

**Comment by lmignon** [comment]
@max3903 we must revert this PR. The code is not right. The definition of the cmis_folder field is wrong. Indeed 2 methods are specified for `create_name_get` and `create_parent_get` but doesn't exist. https://github.com/OCA/connector-cmis/pull/27/files#diff-cad2315bcf26f0f7822f6e7833dc3fe1R13
Moreover, the module depends of an non OCA addon. If I'm not wrong, it's not allowed by OCA.

Sorry for my late review....

**Comment by pedrobaeza** [comment]
Yeah, you can't depend on alfodoo that is outside OCA, although it comes from a well known OCA contributor. Reverting it.

**Comment by pedrobaeza** [comment]
Reverted at bddfe8e180f815f4944edc87b187ce82de2bf47c

**Comment by max3903** [comment]
@lmignon Any plan to move Alfodoo into the OCA?

**Comment by max3903** [comment]
Moved to https://github.com/ursais/osi-addons/pull/2

**Pull request merged**